src/dmqproto/node/neo/request/Consume.d

```diff
@@ -1004,3 +1004,470 @@ public abstract scope class ConsumeProtocol_v2
+
+        /// Aquired buffer in which the most recently popped value is stored.
+        private void[]* value_buffer;
+
+        /// Aquired buffer in which the batch of records to send is stored.
+        private void[]* record_batch;
+
+        /// Constructor, starts the fiber.
+        private this ( )
+        {
+            this.value_buffer = this.outer.resources.getVoidBuffer();
+            this.record_batch = this.outer.resources.getVoidBuffer();
+            this.fiber = this.outer.resources.getFiber(&this.fiberMethod);
+            this.fiber.start();
+        }
+
+        /// Fiber main method.
+        private void fiberMethod ( )
+        {
+            // The loop to send records to the client. The `break` condition for
+            // this loop is that the client signals`Stop`.
+            SendAndWaitLoop: while (true)
+            {
+                // Pop records from the storage and add them to the batch of
+                // records. If the batch is full then send it and wait for a
+                // feedback i.e. `Continue` or `Stop` message from the client.
+                while (this.outer.getNextValue(*this.value_buffer))
+                {
+                    size_t record_length = (*this.value_buffer).length;
+                    (*this.record_batch) ~= (&record_length)[0 .. 1];
+                    (*this.record_batch) ~= *this.value_buffer;
+
+                    if ((*this.record_batch).length >= min_batch_length)
+                    {
+                        this.sendBatch();
+                        if (!this.waitForFeedback())
+                            break SendAndWaitLoop; // Received a `Stop` message.
+                    }
+                }
+
+                // All records have been popped from the storage. Wait for
+                //  - a record to be pushed in the storage or
+                //  - a `Flush` signal (if there are records in the batch) or
+                //  - a `Stop` message from the client.
+                this.outer.resume_fiber_on_push = true;
+                this.outer.resume_fiber_on_flush = !!(*this.record_batch).length;
+
+                auto sigcode = this.outer.request_event_dispatcher.nextEvent(
+                    this.fiber,
+                    Signal(SignalCode.Pushed), Signal(SignalCode.Flush),
+                    Signal(SignalCode.Stop)
+                ).signal.code;
+
+                this.outer.resume_fiber_on_push = false;
+                this.outer.resume_fiber_on_flush = false;
+
+                switch (sigcode)
+                {
+                    case SignalCode.Pushed:
+                        // Continue `SendAndWaitLoop` to pop and send the record
+                        break;
+
+                    case SignalCode.Flush:
+                        // Send the imcomplete batch and wait for a feedback
+                        // i.e. `Continue` or `Stop` message from the node.
+                        this.sendBatch();
```

Comment:
Should do nothing, if the batch is empty. (There's no point sending an empty batch over the wire.)

Reply:
If the batch is empty then `flushBatch` won't raise this signal because the `resume_fiber_on_flush` flag is `false`.